### PR TITLE
devmapper: udev sync is a requirement

### LINKED ISSUE #4036: docker fails to mount the block device for the container on devicemapper
When running something like `for i in {0..100}; do docker run busybox echo test; done` with Docker running on devicemapper, errors are thrown and containers fail to run:

```
2014/02/10 9:48:42 Error: start: Cannot start container 56bee8c4da5bd5641fc42405c742083b418ca14ddfb4a3e632955e236e23c284: Error getting container 56bee8c4da5bd5641fc42405c742083b418ca14ddfb4a3e632955e236e23c284 from driver devicemapper: Error mounting '/dev/mapper/docker-8:1-4980769-56bee8c4da5bd5641fc42405c742083b418ca14ddfb4a3e632955e236e23c284' on '/var/lib/docker/devicemapper/mnt/56bee8c4da5bd5641fc42405c742083b418ca14ddfb4a3e632955e236e23c284': no such file or directory
2014/02/10 9:48:42 Error: start: Cannot start container b90b4385778142aab5251846460008e5c4eb9fe1e7ec82f07d06f1de823bd914: Error getting container b90b4385778142aab5251846460008e5c4eb9fe1e7ec82f07d06f1de823bd914 from driver devicemapper: Error mounting '/dev/mapper/docker-8:1-4980769-b90b4385778142aab5251846460008e5c4eb9fe1e7ec82f07d06f1de823bd914' on '/var/lib/docker/devicemapper/mnt/b90b4385778142aab5251846460008e5c4eb9fe1e7ec82f07d06f1de823bd914': no such file or directory
2014/02/10 9:48:43 Error: start: Cannot start container ca53b3b21c92ffb17ad15c1088be293260ea240abdf25db7e5aadc11517cf93c: Error getting container ca53b3b21c92ffb17ad15c1088be293260ea240abdf25db7e5aadc11517cf93c from driver devicemapper: Error mounting '/dev/mapper/docker-8:1-4980769-ca53b3b21c92ffb17ad15c1088be293260ea240abdf25db7e5aadc11517cf93c' on '/var/lib/docker/devicemapper/mnt/ca53b3b21c92ffb17ad15c1088be293260ea240abdf25db7e5aadc11517cf93c': no such file or directory
test
2014/02/10 9:48:43 Error: start: Cannot start container 1e1e06044711e73cede8ede10547de7e270c33fac7ad5e60a8cb23246950adf3: Error getting container 1e1e06044711e73cede8ede10547de7e270c33fac7ad5e60a8cb23246950adf3 from driver devicemapper: Error mounting '/dev/mapper/docker-8:1-4980769-1e1e06044711e73cede8ede10547de7e270c33fac7ad5e60a8cb23246950adf3' on '/var/lib/docker/devicemapper/mnt/1e1e06044711e73cede8ede10547de7e270c33fac7ad5e60a8cb23246950adf3': no such file or directory
```
Fedora 20 with kernel 3.12.9 doesn't seem to be affected.

kernel version, distribution, docker info and docker version:
```
3.11.0-15-generic #25~precise1-Ubuntu SMP Thu Jan 30 17:39:31 UTC 2014 x86_64 x86_64
Ubuntu 12.04.4
 docker info
Containers: 101
Images: 44
Driver: devicemapper
 Pool Name: docker-8:1-4980769-pool
 Data file: /var/lib/docker/devicemapper/devicemapper/data
 Metadata file: /var/lib/docker/devicemapper/devicemapper/metadata
 Data Space Used: 3234.9 Mb
 Data Space Total: 102400.0 Mb
 Metadata Space Used: 6.9 Mb
 Metadata Space Total: 2048.0 Mb

Client version: 0.8.0-dev
Go version (client): go1.2
Git commit (client): 695719b
Server version: 0.8.0-dev
Git commit (server): 695719b
Go version (server): go1.2
Last stable version: 0.8.0
```

The Docker binary is actually master with PR #4017 merged.

/cc @alexlarsson

### PR DESCRIPTION
**Pull request by vbatts**
Rather than disabling devmapper for static builds, this will get more at the root. Udev sync is a prerequisite for devicemapper. That sync is false on static builds as well.

- closes  https://github.com/docker/docker/issues/4036
- closes #10664

### CODE REVIEW

**Comment by vbatts** [comment]
perhaps this also needs a storage option parameter like `dm.use_regardless_of_udev_sync_i_know_what_i_am_asking_for=true`
for the folks that are using this on ubuntu and being affected, but we can't help due the version of their udev and libdevmapper...

**Comment by jfrazelle** [comment]
we run tests on a fedora machine on every push to master we could make sure its tested there

**Comment by unclejack** [comment]
@vbatts I think we need a special flag to allow this to be bypassed so that it can be tested.

**Comment by vbatts** [comment]
Will do

**Comment by vbatts** [comment]
@jfrazelle @unclejack @rhvgoyal added a flag

**Comment by rhvgoyal** [comment]
@vbatts 

Few more suggestions for flag name.

"dm.ignore_udev_sync_check"
"dm.override_udev_sync_check"

**Comment by rhvgoyal** [comment]
@vbatts 

Can we also modify tests to use above flag by default?

**Comment by vbatts** [comment]
perhaps 'override...', because we are still checking it, so we can log in the daemon that it's an issue.

**Comment by rhvgoyal** [comment]
@vbatts 

"dm.override_udev_sync_check" sounds good.

**Comment by vbatts** [comment]
@rhvgoyal updated PTAL

**Comment by vbatts** [comment]
and added a default for the tests to override the sync requirement

**Comment by rhvgoyal** [comment]
@vbatts 

This patch looks good to me. I have a general question though about tests. Generic tests in docker are run with what graphdriver?

For example, say, I want to run integration tests and one of the test is whether container creation is successful or not. Will I run same test against all graph driver or there is a default graph driver which will always be used or something else?

**Comment by vbatts** [comment]
vfs is the driver the Docker-in-docker daemon is running for tests.

**Comment by rhvgoyal** [comment]
Ok, thanks.

LGTM.

**Comment by rhvgoyal** [comment]
@vbatts 

Thinking more about it. Actually this change is not upgrade proof. Those who are using device mapper and when they upgrade docker, suddenly docker daemon will stop running and that's not good.

**Comment by vbatts** [comment]
that's why the storage-opt flag is there. It's not supported in the way
they're running it presently.
if with the upgrade, they are then required to add the flag, it will be an
explicit indication that they're running it in an unsupported way, but
still have the ability to do so. Even if for migrating to a supported
setup, or switching to a docker binary that supports udev sync.

**Comment by rhvgoyal** [comment]
Generally I am not a big fan of such approach. Typically first we move users to a different storage driver slowly before switch is made.

Given that so many people are complaining, clearly lot of people are using it. They will be upset when upgraded docker suddenly does not start.

In kernel, we have a deprecated file where we put a feature which we want to deprecate. Send loud warnings if user is using those features in an effort to migrate them to alternate means. And once we are reasonably sure that most of the user's have migrated, we make the switch.

Personally, I would rather continue to give loud warnings and tell users to move to aufs or overlayfs or move to dynamic binary when they complain, instead of breaking working setup.  Yes, flag is there, but it is very annoying to figure out that one needs to pass a new flag before one can proceed.

All I am saying that existing users might not like this change.

**Comment by unclejack** [comment]
Perhaps calling it something like `unsafe_override_udev_sync_check` would send a very clear message that this isn't safe, nor supported. Whatever it's called, we probably want to have something like `unsafe` or `unsupported` in the flag itself to make it clear this voids the warranty.

**Comment by rhvgoyal** [comment]
@unclejack 

I don't think that naming of flag is an issue here. Problem is that docker 1.5 was working in a setup and same setup will suddenly stop working on 1.6. And that's a problem.

We agree that whatever was working in 1.5 is not supportable, but that ship has sailed and people are already using it. I think instead of failing it completely, a better approach is to just give loud warnings that is is not supported and will go away in future, write documentation and point people to documentation.

And once we are reasonably sure that people have migrated to different solutions, then make this switch, IMHO.

**Comment by unclejack** [comment]
@rhvgoyal You're right, but those users also need to know as soon as possible that they're exposing themselves and any potential clients they may have to data corruption/loss on the Docker daemons where they're running such a setup.

We need to provide apt and yum repositories which are distro specific in order to be able to provide dynamically linked binaires. That's a requirement for getting this merged, IMO.

**Comment by vbatts** [comment]
should we inverse this flag for now, so that we can make it available to
the pedantic. and once the requirements are met, we can switch this default
to fail support when udev sync is not supported?

**Comment by rhvgoyal** [comment]
@vbatts 

By inversing the flag you mean provide a flag to fail initialization of devicemapper if udev sync is not supported. And without presence of flag, you will allow devicemapper to initialize?

If yes, I am not sure how useful it is. If somebody knows this issue then flag does not help him. And if somebody does not know the issue, then he does not know the flag either and then its of no use to him. 

I think right now we seem to have two choices only and we need to pick our poison.

- Just continue to give loud warnings and write documentation to point to users in right direction.

- Implement this flag and in error message tell user why daemon has failed and if they want to 
  override this behavior how to do it. That way, it will atleast make it easy for users to find this new
  flag. So this will still be annoying but I guess less painful. One will have to still write documentation
   to explain what are the alternatives.

**Comment by vbatts** [comment]
@rhvgoyal will do

**Comment by vbatts** [comment]
rebased and updated. 
PTAL

**Comment by rhvgoyal** [comment]
LGTM

File: daemon/graphdriver/devmapper/deviceset.go [hunk 1]
```diff
@@ -963,9 +965,11 @@ func (devices *DeviceSet) initDevmapper(doInit bool) error {
 
 	// https://github.com/docker/docker/issues/4036
 	if supported := devicemapper.UdevSetSyncSupport(true); !supported {
-		logrus.Warnf("Udev sync is not supported. This will lead to unexpected behavior, data loss and errors")
+		logrus.Errorf("Udev sync is not supported. This will lead to unexpected behavior, data loss and errors. For more information, see http://docs.docker.com/reference/commandline/cli/#daemon-storage-driver-option")
```
Comment (tiborvass) [inline daemon/graphdriver/devmapper/deviceset.go]: `https`

Comment (vbatts) [inline daemon/graphdriver/devmapper/deviceset.go]: ... k
Though this should force a redirect if thats the case, because this is
where google sent me ...

**Comment by vbatts** [comment]
fixed

**Comment by tiborvass** [comment]
LGTM

Ping @unclejack

**Comment by tiborvass** [comment]
@unclejack once you're good, please move to docs-review, thanks!

**Comment by unclejack** [comment]
LGTM

**Comment by tiborvass** [comment]
Ping @fredlf @moxiegirl @SvenDowideit

File: daemon/graphdriver/devmapper/README.md [hunk 2]
```diff
@@ -216,3 +216,26 @@ Here is the list of supported options:
     Example use:
 
     ``docker -d --storage-opt dm.blkdiscard=false``
+
```
Comment (moxiegirl) [inline daemon/graphdriver/devmapper/README.md]: There were some awkward constructions that made this hard to parse and obscured the behavior.  The Notes and Warning on top of each other indicate structural issues. Hopefully, I cleared this up with some input from @tiborvass  (ty!).

Markdown: https://gist.github.com/moxiegirl/ade0634692f7423bf474

---->

*  `dm.override_udev_sync_check`

Overrides the kernel's `udev` synchronization checks. `udev` is the device
manager for the Linux kernel. By default, this value is `false` and the kernel
does not support override `udev` synchronization checks.  

Under the `false` setting, race conditions occur between the`devicemapper` and `udev` during create and cleanup. The race conditions result in errors and the
`devicemapper` fails. (For information on these failures, see
[docker#4036](https://github.com/docker/docker/issues/4036))

To allow the `docker` daemon to function properly with the `devicemapper`
driver, set `dm.override_udev_sync_check` to true:  

	$ docker -d --storage-opt dm.override_udev_sync_check=true
	
When this value is `true`, the  `devicemapper` continues and simply warns you the errors are happening 

**Note**: The ideal is to pursue a `docker` daemon and environment that does
support synchronizing with Udev. For further discussion on this topic, see 
[docker#4036](https://github.com/docker/docker/issues/4036).  Otherwise, set this flag for migrating existing daemons to a daemon
with a supported environment.

File: docs/sources/reference/commandline/cli.md [hunk 3]
```diff
@@ -370,6 +370,28 @@ Currently supported options are:
 
         $ docker -d --storage-opt dm.blkdiscard=false
 
+ *  `dm.override_udev_sync_check`
```
Comment (moxiegirl) [inline docs/sources/reference/commandline/cli.md]: Same as above.

**Comment by vbatts** [comment]
@moxiegirl I will update the wording. I see that it is not clear enough. :-)

**Comment by vbatts** [comment]
docs updated PTAL

**Comment by moxiegirl** [comment]
@vbatts Thanks for the update. You missed a few ticks around `udev` and `devicemapper` line 381 --- but I can clean those up later.  LGTM

**Comment by moxiegirl** [comment]
ping @fredlf @SvenDowideit @jamtur01

File: daemon/graphdriver/devmapper/README.md [hunk 4]
```diff
@@ -216,3 +216,39 @@ Here is the list of supported options:
...
+ *  `dm.override_udev_sync_check`
+
+    Overrides the `udev` synchronization checks between devicemapper and udev.
```
Comment (jamtur01) [inline daemon/graphdriver/devmapper/README.md]: \`devicemapper\` \`udev\`

File: daemon/graphdriver/devmapper/README.md [hunk 5]
```diff
@@ -216,3 +216,39 @@ Here is the list of supported options:
...
+        $ docker info
+	[...]
+	 Udev Sync Supported: true
+	[...]
+
+    When `udev` sync is supported, then devicemapper and udev can coordinate
+    the activation and deactivation of devices for containers.
+
+    When `udev` sync is `false`, a race conditions occurs between
+    the`devicemapper` and `udev` during create and cleanup. The race conditions
```
Comment (jamtur01) [inline daemon/graphdriver/devmapper/README.md]: condition

File: daemon/graphdriver/devmapper/README.md [hunk 6]
```diff
@@ -216,3 +216,39 @@ Here is the list of supported options:
...
+    result in errors and failures. (For information on these failures, see
```
Comment (jamtur01) [inline daemon/graphdriver/devmapper/README.md]: results

File: daemon/graphdriver/devmapper/README.md [hunk 7]
```diff
@@ -216,3 +216,39 @@ Here is the list of supported options:
...
+
+    To allow the `docker` daemon to start, regardless of `udev` sync not being
+    supported, set `dm.override_udev_sync_check` to true:
+
+        $ docker -d --storage-opt dm.override_udev_sync_check=true
+
+    When this value is `true`, the  `devicemapper` continues and simply warns
+    you the errors are happening.
+
+    **Note**: The ideal is to pursue a `docker` daemon and environment that
```
Comment (jamtur01) [inline daemon/graphdriver/devmapper/README.md]: Is this formatted as a proper note?

File: daemon/graphdriver/devmapper/README.md [hunk 8]
```diff
@@ -216,3 +216,39 @@ Here is the list of supported options:
...
+    does support synchronizing with Udev. For further discussion on this topic,
```
Comment (jamtur01) [inline daemon/graphdriver/devmapper/README.md]: Consistency of use - udev/Udev.

File: daemon/graphdriver/devmapper/README.md [hunk 9]
```diff
@@ -216,3 +216,39 @@ Here is the list of supported options:
...
+    see [docker#4036](https://github.com/docker/docker/issues/4036).
+    Otherwise, set this flag for migrating existing daemons to a daemon with a
```
Comment (jamtur01) [inline daemon/graphdriver/devmapper/README.md]: migrating existing Docker daemons ...

File: docs/sources/reference/commandline/cli.md [hunk 10]
```diff
@@ -370,6 +370,41 @@ Currently supported options are:
...
+
+    Overrides the `udev` synchronization checks between devicemapper and udev.
```
Comment (jamtur01) [inline docs/sources/reference/commandline/cli.md]: Comments as above.

**Comment by jamtur01** [comment]
@vbatts Did you push an update? Perhaps wait until I get a chance to finish the review? :) Makes it hard to go back and see what's changed.

**Comment by vbatts** [comment]
@jamtur01 sorry! i thought i saw them slow down, so i pushed :-\

**Comment by jamtur01** [comment]
No worries - love the lightning fast fixes! :)

File: docs/sources/reference/commandline/cli.md [hunk 11]
```diff
@@ -370,6 +370,41 @@ Currently supported options are:
...
+
+        $ docker info
+	[...]
+	 Udev Sync Supported: true
+	[...]
+
+    When `udev` sync is supported, then devicemapper and udev can coordinate
+    the activation and deactivation of devices for containers.
+
+    When `udev` sync is `false`, a race conditions occurs between
```
Comment (tiborvass) [inline docs/sources/reference/commandline/cli.md]: `udev` sync is not supported 

or

`udev` sync support is `false`

Ping @moxiegirl @jamtur01

File: daemon/graphdriver/devmapper/README.md (hunk 7 above)
Comment (moxiegirl) [inline daemon/graphdriver/devmapper/README.md]: No it isn't. That is my fault. @vbatts I hate to pepper ya with this formatting. If you have time to update. I've fixed the gist:

https://gist.github.com/moxiegirl/ade0634692f7423bf474

File: docs/sources/reference/commandline/cli.md (hunk 11 above)
Comment (vbatts) [inline docs/sources/reference/commandline/cli.md]: ah, perhaps "supported" would be good there.

File: daemon/graphdriver/devmapper/README.md (hunk 7 above)
Comment (vbatts) [inline daemon/graphdriver/devmapper/README.md]: ah. let me get that.

**Comment by vbatts** [comment]
Notes formatting updated. PTAL

**Comment by moxiegirl** [comment]
@jamtur01 ping back...let us know if you are satisfied. :D

**Comment by jamtur01** [comment]
Docs LGTM
